File: lib/lastfm.dart
import 'dart:convert';

import 'package:crypto/crypto.dart';
import 'package:finale/env.dart';
import 'package:finale/types/generic.dart';
import 'package:finale/types/lalbum.dart';
import 'package:finale/types/lartist.dart';
import 'package:finale/types/lcommon.dart';
import 'package:finale/types/ltrack.dart';
import 'package:finale/types/luser.dart';
import 'package:http/http.dart';
import 'package:http_throttle/http_throttle.dart';
import 'package:shared_preferences/shared_preferences.dart';

enum RequestVerb { get, post }

final _client = ThrottleClient(10);

Uri _buildUri(String method, Map<String, dynamic> data) {
  final allData = {
    ...data.map((key, value) => MapEntry(key, value.toString())),
    'api_key': apiKey,
    'method': method,
  };

  final hash = (allData.keys.toList()..sort())
          .map((key) => '$key${allData[key]}')
          .join() +
      apiSecret;
  final signature = md5.convert(utf8.encode(hash));
  allData['api_sig'] = signature.toString();
  allData['format'] = 'json';

  return Uri(
      scheme: 'https',
      host: 'ws.audioscrobbler.com',
      path: '2.0',
      queryParameters: allData);
}

Future<Map<String, dynamic>> _doRequest(
    String method, Map<String, dynamic> data,
    {RequestVerb verb = RequestVerb.get}) async {
  final uri = _buildUri(method, data);

  final response = verb == RequestVerb.get
      ? await _client.get(uri)
      : await _client.post(uri);

  if (response.statusCode == 200) {
    return json.decode(utf8.decode(response.bodyBytes));
  } else if (response.statusCode == 400) {
    final error = LError.fromJson(json.decode(utf8.decode(response.bodyBytes)));
    throw LException(error.code, error.message);
  } else {
    throw Exception('Could not do request $method');
  }
}

abstract class PagedLastfmRequest<T> {
  Future<List<T>> doRequest(int limit, int page, {String period});
}

class GetRecentTracksRequest
    extends PagedLastfmRequest<LRecentTracksResponseTrack> {
  String username;

  GetRecentTracksRequest(this.username);

  @override
  doRequest(int limit, int page, {String period}) async {
    final rawResponse = await _doRequest('user.getRecentTracks',
        {'user': username, 'limit': limit, 'page': page});
    final tracks =
        LRecentTracksResponseRecentTracks.fromJson(rawResponse['recenttracks'])
            .tracks;

    // For some reason, this endpoint always returns the currently-playing
    // song regardless of which page you request.
    if (page != 1 && tracks.isNotEmpty && tracks.first.date == null) {
      tracks.removeAt(0);
    }

    return tracks;
  }
}

class GetTopArtistsRequest
    extends PagedLastfmRequest<LTopArtistsResponseArtist> {
  String username;

  GetTopArtistsRequest(this.username);

  @override
  doRequest(int limit, int page, {String period}) async {
    final rawResponse = await _doRequest('user.getTopArtists',
        {'user': username, 'limit': limit, 'page': page, 'period': period});
    return LTopArtistsResponseTopArtists.fromJson(rawResponse['topartists'])
        .artists;
  }
}

class GetTopAlbumsRequest extends PagedLastfmRequest<LTopAlbumsResponseAlbum> {
  String username;

  GetTopAlbumsRequest(this.username);

  @override
  doRequest(int limit, int page, {String period}) async {
    final rawResponse = await _doRequest('user.getTopAlbums',
        {'user': username, 'limit': limit, 'page': page, 'period': period});
    return LTopAlbumsResponseTopAlbums.fromJson(rawResponse['topalbums'])
        .albums;
  }
}

class GetTopTracksRequest extends PagedLastfmRequest<LTopTracksResponseTrack> {
  String username;

  GetTopTracksRequest(this.username);

  @override
  doRequest(int limit, int page, {String period}) async {
    final rawResponse = await _doRequest('user.getTopTracks',
        {'user': username, 'limit': limit, 'page': page, 'period': period});
    return LTopTracksResponseTopTracks.fromJson(rawResponse['toptracks'])
        .tracks;
  }
}

class GetFriendsRequest extends PagedLastfmRequest<LUser> {
  String username;

  GetFriendsRequest(this.username);

  @override
  doRequest(int limit, int page, {String period}) async {
    final rawResponse = await _doRequest(
        'user.getFriends', {'user': username, 'limit': limit, 'page': page});
    return LUserFriendsResponse.fromJson(rawResponse['friends']).friends;
  }
}

class SearchTracksRequest extends PagedLastfmRequest<LTrackMatch> {
  String query;

  SearchTracksRequest(this.query);

  @override
  Future<List<LTrackMatch>> doRequest(int limit, int page,
      {String period}) async {
    final rawResponse = await _doRequest(
        'track.search', {'track': query, 'limit': limit, 'page': page});
    return LTrackSearchResponse.fromJson(rawResponse['results']['trackmatches'])
        .tracks;
  }
}

class SearchArtistsRequest extends PagedLastfmRequest<LArtistMatch> {
  String query;

  SearchArtistsRequest(this.query);

  @override
  Future<List<LArtistMatch>> doRequest(int limit, int page,
      {String period}) async {
    final rawResponse = await _doRequest(
        'artist.search', {'artist': query, 'limit': limit, 'page': page});
    return LArtistSearchResponse.fromJson(
            rawResponse['results']['artistmatches'])
        .artists;
  }
}

class SearchAlbumsRequest extends PagedLastfmRequest<LAlbumMatch> {
  String query;

  SearchAlbumsRequest(this.query);

  @override
  Future<List<LAlbumMatch>> doRequest(int limit, int page,
      {String period}) async {
    final rawResponse = await _doRequest(
        'album.search', {'album': query, 'limit': limit, 'page': page});
    return LAlbumSearchResponse.fromJson(rawResponse['results']['albummatches'])
        .albums;
  }
}

class ArtistGetTopAlbumsRequest extends PagedLastfmRequest<LArtistTopAlbum> {
  String artist;

  ArtistGetTopAlbumsRequest(this.artist);

  @override
  Future<List<LArtistTopAlbum>> doRequest(int limit, int page,
      {String period}) async {
    final rawResponse = await _doRequest('artist.getTopAlbums',
        {'artist': artist, 'limit': limit, 'page': page});
    return LArtistGetTopAlbumsResponse.fromJson(rawResponse['topalbums'])
        .albums;
  }
}

class ArtistGetTopTracksRequest extends PagedLastfmRequest<LArtistTopTrack> {
  String artist;

  ArtistGetTopTracksRequest(this.artist);

  @override
  Future<List<LArtistTopTrack>> doRequest(int limit, int page,
      {String period}) async {
    final rawResponse = await _doRequest('artist.getTopTracks',
        {'artist': artist, 'limit': limit, 'page': page});
    return LArtistGetTopTracksResponse.fromJson(rawResponse['toptracks'])
        .tracks;
  }
}

class Lastfm {
  static Future<Response> get(String url) => _client.get(url);

  static Future<LAuthenticationResponseSession> authenticate(
      String token) async {
    final rawResponse = await _doRequest('auth.getSession', {'token': token},
        verb: RequestVerb.post);
    return LAuthenticationResponseSession.fromJson(rawResponse['session']);
  }

  static Future<LUser> getUser(String username) async {
    final rawResponse = await _doRequest('user.getInfo', {'user': username});
    return LUser.fromJson(rawResponse['user']);
  }

  static Future<LTrack> getTrack(BasicTrack track) async {
    final username = (await SharedPreferences.getInstance()).getString('name');

    final rawResponse = await _doRequest('track.getInfo',
        {'track': track.name, 'artist': track.artist, 'username': username});
    return LTrack.fromJson(rawResponse['track']);
  }

  static Future<LAlbum> getAlbum(BasicAlbum album) async {
    final username = (await SharedPreferences.getInstance()).getString('name');

    final rawResponse = await _doRequest('album.getInfo', {
      'album': album.name,
      'artist': album.artist.name,
      'username': username
    });
    return LAlbum.fromJson(rawResponse['album']);
  }

  static Future<LArtist> getArtist(BasicArtist artist) async {
    final username = (await SharedPreferences.getInstance()).getString('name');

    final rawResponse = await _doRequest(
        'artist.getInfo', {'artist': artist.name, 'username': username});
    return LArtist.fromJson(rawResponse['artist']);
  }

  static Future<LUserWeeklyChartList> getWeeklyChartList(LUser user) async {
    final rawResponse =
        await _doRequest('user.getWeeklyChartList', {'user': user.name});
    return LUserWeeklyChartList.fromJson(rawResponse['weeklychartlist']);
  }

  static Future<LUserWeeklyTrackChart> getWeeklyTrackChart(
      LUser user, LUserWeeklyChart chart) async {
    final rawResponse = await _doRequest('user.getWeeklyTrackChart', {
      'user': user.name,
      'from': chart.from,
      'to': chart.to,
    });
    return LUserWeeklyTrackChart.fromJson(rawResponse['weeklytrackchart']);
  }

  static Future<LUserWeeklyAlbumChart> getWeeklyAlbumChart(
      LUser user, LUserWeeklyChart chart) async {
    final rawResponse = await _doRequest('user.getWeeklyAlbumChart', {
      'user': user.name,
      'from': chart.from,
      'to': chart.to,
    });
    return LUserWeeklyAlbumChart.fromJson(rawResponse['weeklyalbumchart']);
  }

  static Future<LUserWeeklyArtistChart> getWeeklyArtistChart(
      LUser user, LUserWeeklyChart chart) async {
    final rawResponse = await _doRequest('user.getWeeklyArtistChart', {
      'user': user.name,
      'from': chart.from,
      'to': chart.to,
    });
    return LUserWeeklyArtistChart.fromJson(rawResponse['weeklyartistchart']);
  }

  static Future<int> getNumArtists(String username) async {
    final rawResponse = await _doRequest('user.getTopArtists',
        {'user': username, 'period': 'overall', 'limit': '1', 'page': '1'});
    return LTopArtistsResponseTopArtists.fromJson(rawResponse['topartists'])
        .attr
        .total;
  }

  static Future<int> getNumAlbums(String username) async {
    final rawResponse = await _doRequest('user.getTopAlbums',
        {'user': username, 'period': 'overall', 'limit': '1', 'page': '1'});
    return LTopAlbumsResponseTopAlbums.fromJson(rawResponse['topalbums'])
        .attr
        .total;
  }

  static Future<int> getNumTracks(String username) async {
    final rawResponse = await _doRequest('user.getTopTracks',
        {'user': username, 'period': 'overall', 'limit': '1', 'page': '1'});
    return LTopTracksResponseTopTracks.fromJson(rawResponse['toptracks'])
        .attr
        .total;
  }

  static Future<List<LTopArtistsResponseArtist>> getGlobalTopArtists(
      int limit) async {
    final rawResponse =
        await _doRequest('chart.getTopArtists', {'limit': limit, 'page': 1});
    return LChartTopArtists.fromJson(rawResponse['artists']).artists;
  }

  static Future<LScrobbleResponseScrobblesAttr> scrobble(
      List<BasicTrack> tracks, List<DateTime> timestamps) async {
    final Map<String, dynamic> data = {};
    data['sk'] = (await SharedPreferences.getInstance()).getString('key');

    tracks.asMap().forEach((i, track) {
      if (track.album.isNotEmpty) {
        data['album[$i]'] = track.album;
      }

      data['artist[$i]'] = track.artist;
      data['track[$i]'] = track.name;
      data['timestamp[$i]'] = timestamps[i].millisecondsSinceEpoch ~/ 1000;
    });

    final rawResponse =
        await _doRequest('track.scrobble', data, verb: RequestVerb.post);
    return LScrobbleResponseScrobblesAttr.fromJson(
        rawResponse['scrobbles']['@attr']);
  }

  /// Loves or unloves a track. If [love] is true, the track will be loved;
  /// otherwise, it will be unloved.
  static Future<bool> love(FullTrack track, bool love) async {
    await _doRequest(
        love ? 'track.love' : 'track.unlove',
        {
          'track': track.name,
          'artist': track.artist.name,
          'sk': (await SharedPreferences.getInstance()).getString('key')
        },
        verb: RequestVerb.post);
    return true;
  }
}

class LException implements Exception {
  int code;
  String message;

  LException(this.code, this.message);

  @override
  String toString() {
    return 'Error $code: $message';
  }
}

File: ios/FinaleWidget/Lastfm.swift
import CryptoKit
import Foundation

@available(iOS 13.0, *)
class Lastfm {
    private static func buildUrl(_ method: String, _ data: [String: Any]) -> URL {
        var allData = data.merging(["api_key": Env.apiKey, "method": method]) { (a, _) in return a }
        let hash = allData.keys.sorted().map({"\($0)\(allData[$0]!)"}).joined() + Env.apiSecret
        let signature = Insecure.MD5.hash(data: hash.data(using: .utf8)!).map { String(format: "%02hhx", $0) }.joined()
        allData["api_sig"] = signature
        allData["format"] = "json"
        
        var components = URLComponents()
        components.scheme = "https"
        components.host = "ws.audioscrobbler.com"
        components.path = "/2.0"
        components.queryItems = allData.map({URLQueryItem(name: $0, value: String(describing: $1))})
        return components.url!
    }
    
    fileprivate static func doRequest<T : Codable>(method: String, data: [String: Any], post: Bool = false, callback: @escaping (T?) -> Void) {
        let url = buildUrl(method, data)
        var request = URLRequest(url: url)
        
        if post {
            request.httpMethod = "post"
        }
        
        let task = URLSession.shared.dataTask(with: request) { data, urlResponse, err in
            if let data = data {
                let decoder = JSONDecoder()
                decoder.dateDecodingStrategy = .secondsSince1970
                
                do {
                    let obj = try decoder.decode(T.self, from: data)
                    callback(obj)
                }
                
                catch {
                    print(error)
                    callback(nil)
                }
            }
        }
        
        task.resume()
    }
    
    static func getFullTrack(from simpleTrack: LTopTracksResponseTrack, callback: @escaping (LTrack?) -> Void) {
        doRequest(method: "track.getInfo", data: ["track": simpleTrack.name, "artist": simpleTrack.artist.name]) { (trackResponse: LTrackInfoResponse?) in
            callback(trackResponse?.track)
        }
    }
}

@available(iOS 13.0, *)
class GetRecentTracksRequest: LastfmApiRequest {
    var username: String
    var period: Period
    
    init(username: String, period: Period) {
        self.username = username
        self.period = period
    }
    
    func doRequest(limit: Int, page: Int, callback: @escaping (LRecentTracksResponse?) -> Void) {
        Lastfm.doRequest(method: "user.getRecentTracks", data: ["user": username, "period": period.apiValue, "limit": limit, "page": page], callback: callback)
    }
}

@available(iOS 13.0, *)
class GetTopTracksRequest: LastfmApiRequest {
    var username: String
    var period: Period
    
    init(username: String, period: Period) {
        self.username = username
        self.period = period
    }
    
    func doRequest(limit: Int, page: Int, callback: @escaping (LTopTracksResponse?) -> Void) {
        Lastfm.doRequest(method: "user.getTopTracks", data: ["user": username, "period": period.apiValue, "limit": limit, "page": page], callback: callback)
    }
}

@available(iOS 13.0, *)
class GetTopArtistsRequest: LastfmApiRequest {
    var username: String
    var period: Period
    
    init(username: String, period: Period) {
        self.username = username
        self.period = period
    }
    
    func doRequest(limit: Int, page: Int, callback: @escaping (LTopArtistsResponse?) -> Void) {
        Lastfm.doRequest(method: "user.getTopArtists", data: ["user": username, "period": period.apiValue, "limit": limit, "page": page], callback: callback)
    }
}

@available(iOS 13.0, *)
class GetTopAlbumsRequest: LastfmApiRequest {
    var username: String
    var period: Period
    
    init(username: String, period: Period) {
        self.username = username
        self.period = period
    }
    
    func doRequest(limit: Int, page: Int, callback: @escaping (LTopAlbumsResponse?) -> Void) {
        Lastfm.doRequest(method: "user.getTopAlbums", data: ["user": username, "period": period.apiValue, "limit": limit, "page": page], callback: callback)
    }
}

protocol LastfmApiRequest {
    associatedtype ResponseWrapper: LastfmApiResponseWrapper
    
    func doRequest(limit: Int, page: Int, callback: @escaping (ResponseWrapper?) -> Void) -> Void
}

extension LastfmApiRequest {
    func getEntities(limit: Int, page: Int, callback: @escaping ([ResponseWrapper.Response.E]?) -> Void) -> Void {
        doRequest(limit: limit, page: page) { wrapper in
            callback(wrapper?.response.entities)
        }
    }
    
    func getTotalCount(callback: @escaping (Int?) -> Void) -> Void {
        doRequest(limit: 1, page: 1) { wrapper in
            if let wrapper = wrapper {
                callback(Int(wrapper.response.attr.total))
            } else {
                callback(nil)
            }
        }
    }
}

protocol LastfmApiResponseWrapper : Codable {
    associatedtype Response : LastfmApiResponse
    
    var response: Response { get }
}

protocol LastfmApiResponse : Codable {
    associatedtype E : Entity
    
    var entities: [E] { get }
    var attr: LAttr { get }
}
